File: app/src/main/java/tomaszkopacz/meetbam/presenters/StatsGlobalFragmentPresenter.java
package tomaszkopacz.meetbam.presenters;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import tomaszkopacz.meetbam.activities.StatsGlobalFragment;
import tomaszkopacz.meetbam.model.UserResult;
import tomaszkopacz.meetbam.service.UserAdapter;
import tomaszkopacz.meetbam.service.UserViewHolder;
import tomaszkopacz.meetbam.service.WebService;

/**
 * Created by tomas on 02.03.2018.
 */

public class StatsGlobalFragmentPresenter implements RecyclerViewPresenter{

    private StatsGlobalFragment fragment;
    private WebService mWebService;

    // users list
    private UserAdapter adapter;
    private List<UserResult> results = new ArrayList<>();

    /**
     * Constructor.
     */
    public StatsGlobalFragmentPresenter(StatsGlobalFragment fragment, WebService service){
        this.fragment = fragment;
        this.mWebService = service;

        this.adapter = new UserAdapter(this);
    }

    /**
     * Downloads users and send them to activity.
     */
    public void downloadResults(){
        // remove previous elements
        results.clear();

        // when new elements downloaded
        Call call = mWebService.getGlobalResults();
        call.enqueue(new Callback<List<UserResult>>() {

            @Override
            public void onResponse(Call<List<UserResult>> call, Response<List<UserResult>> response) {

                // assign response objects to list
                results = response.body();

                // send users to activity
                fragment.putResults(adapter);
            }

            @Override
            public void onFailure(Call<List<UserResult>> call, Throwable t) {

            }
        });
    }

    @Override
    public void onItemBoundAtPosition(RecyclerView.ViewHolder holder, int position) {
        UserResult result = results.get(position);
        ((UserViewHolder)holder).getUsername()
                .setText(position + 1 + ". " + result.getName() + " " + result.getSurname());

        ((UserViewHolder)holder).getResult()
                .setText(result.getResult() + " posts");
    }

    @Override
    public void onItemClick(View view) {

    }

    @Override
    public int getItemCount() {
        return results.size();
    }
}

File: app/src/main/java/tomaszkopacz/meetbam/presenter/StatsGlobalFragmentPresenter.kt
package tomaszkopacz.meetbam.presenter

import android.support.v7.widget.RecyclerView
import android.view.View
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import tomaszkopacz.meetbam.R
import tomaszkopacz.meetbam.entity.UserResult
import tomaszkopacz.meetbam.interactor.WebService
import tomaszkopacz.meetbam.service.UserAdapter
import tomaszkopacz.meetbam.view.MainApp
import tomaszkopacz.meetbam.view.StatsGlobalFragment
import java.util.*
import javax.inject.Inject


class StatsGlobalFragmentPresenter(private val fragment: StatsGlobalFragment)
    : RecyclerViewPresenter {

    //service
    @Inject lateinit var webService: WebService

    private val adapter = UserAdapter(this)
    private var results: MutableList<UserResult>? = ArrayList()

    init {
        (fragment.activity!!.application as MainApp)
                .webServiceComponent!!.inject(this)
    }

    fun downloadResults() {
        results!!.clear()

        val call = webService.globalResults
        call.enqueue(object : Callback<List<UserResult>> {

            override fun onResponse(call: Call<List<UserResult>>, response: Response<List<UserResult>>) {

                results = response.body() as MutableList<UserResult>?
                fragment.putResults(adapter)
            }

            override fun onFailure(call: Call<List<UserResult>>, t: Throwable) {

            }
        })
    }

    override fun onItemBoundAtPosition(holder: RecyclerView.ViewHolder, position: Int) {
        val result = results!![position]

        holder as UserAdapter.UserViewHolder
        holder.username.text =
                fragment.getString(R.string.user_at_position,
                        position + 1, result.name, result.surname)

        holder.result.text = fragment.getString(R.string.posts_number, result.result)
    }

    override fun onItemClick(view: View) {

    }

    override fun onLongItemClick(view: View) {

    }

    override fun itemCount(): Int{
        return results!!.size
    }
}
